File: src/bot_python/storage.py
import json
import os
import datetime
from typing import Dict, List, Optional, Tuple

class JSONRecountStorage:
    def __init__(self, file_path: str = "recount_data.json", max_slots: int = 100):
        self.file_path = file_path
        self.max_slots = max_slots
        self.data = self._load_data()

    def _load_data(self) -> dict:
        if os.path.exists(self.file_path):
            try:
                with open(self.file_path, "r", encoding="utf-8") as f:
                    data = json.load(f)
                    if "users" not in data:
                        data["users"] = {}
                    if "slots" not in data:
                        data["slots"] = {}
                    if "check_ins" not in data:
                        data["check_ins"] = {}
                    return data
            except Exception as e:
                print(f"Ошибка при загрузке JSON: {e}. Инициализируем новую базу.")
        
        return {
            "users": {},      # user_id (str) -> {id, username, first_name, last_name, slot, registered_at}
            "slots": {},      # slot_number (str) -> user_id
            "check_ins": {}   # date_str -> list of user_ids (str)
        }

    def _save_data(self):
        try:
            with open(self.file_path, "w", encoding="utf-8") as f:
                json.dump(self.data, f, ensure_ascii=False, indent=2)
        except Exception as e:
            print(f"Ошибка сохранения JSON: {e}")

    def get_user(self, user_id: int) -> Optional[dict]:
        user_id_str = str(user_id)
        return self.data["users"].get(user_id_str)

    def get_all_users(self) -> List[dict]:
        return list(self.data["users"].values())

    def choose_slot(self, user_id: int, username: Optional[str], first_name: str, last_name: Optional[str], slot_number: int) -> Tuple[bool, Optional[str]]:
        if slot_number < 1 or slot_number > self.max_slots:
            return False, f"Номер должен быть от 1 до {self.max_slots}!"

        user_id_str = str(user_id)
        slot_str = str(slot_number)

        # Check if user already has a slot
        user = self.get_user(user_id)
        if user and user.get("slot_number") is not None:
            return False, f"Вы уже выбрали номер — это Пидараз {user['slot_number']}. Изменить его нельзя!"

        # Check if slot already taken
        taken_by = self.data["slots"].get(slot_str)
        if taken_by and taken_by != user_id_str:
            return False, f"Номер {slot_number} уже занят другим пидаразом!"

        # Process reservation
        now_str = datetime.datetime.now().isoformat()
        user_info = {
            "id": user_id,
            "username": username,
            "first_name": first_name,
            "last_name": last_name,
            "slot_number": slot_number,
            "registered_at": now_str
        }

        self.data["users"][user_id_str] = user_info
        self.data["slots"][slot_str] = user_id_str
        self._save_data()

        return True, None

    def record_check_in(self, user_id: int, date_str: str) -> bool:
        """
        Record morning presence check-in. Returns True if recorded successfully, 
        False if user not found, or already checked in today.
        """
        user_id_str = str(user_id)
        user = self.get_user(user_id)
        if not user or user.get("slot_number") is None:
            return False

        if date_str not in self.data["check_ins"]:
            self.data["check_ins"][date_str] = []

        if user_id_str not in self.data["check_ins"][date_str]:
            self.data["check_ins"][date_str].append(user_id_str)
            self._save_data()
            return True

        return False

    def get_check_ins(self, date_str: str) -> List[int]:
        user_id_strs = self.data["check_ins"].get(date_str, [])
        return [int(uid) for uid in user_id_strs]

    def get_slots(self, limit: int = 100) -> List[dict]:
        slots_list = []
        for i in range(1, limit + 1):
            slot_str = str(i)
            user_id_str = self.data["slots"].get(slot_str)
            slots_list.append({
                "number": i,
                "user_id": int(user_id_str) if user_id_str else None
            })
        return slots_list

File: src/bot/storage.ts
import * as fs from 'fs';
import * as path from 'path';
import { RecountState, TeleUser, Slot } from './types';

export interface IBotStorage {
  getUser(userId: number): Promise<TeleUser | null>;
  saveUser(user: TeleUser): Promise<void>;
  chooseSlot(userId: number, username: string | undefined, firstName: string, lastName: string | undefined, slotNumber: number): Promise<{ success: boolean; error?: string }>;
  getSlots(limit: number): Promise<Slot[]>;
  getAllUsers(): Promise<TeleUser[]>;
  recordCheckIn(userId: number, dateStr: string): Promise<boolean>;
  getCheckIns(dateStr: string): Promise<number[]>;
}

export class JsonFileStorage implements IBotStorage {
  private filePath: string;
  private state: RecountState;
  private maxSlots: number;

  constructor(filePath?: string, maxSlots = 100) {
    this.filePath = filePath || path.join(process.cwd(), 'recount-data.json');
    this.maxSlots = maxSlots;
    this.state = this.loadState();
  }

  private loadState(): RecountState {
    try {
      if (fs.existsSync(this.filePath)) {
        const fileContent = fs.readFileSync(this.filePath, 'utf-8');
        const parsed = JSON.parse(fileContent) as RecountState;
        
        // Ensure slots structure is fully populated
        if (!parsed.slots) parsed.slots = {};
        if (!parsed.users) parsed.users = {};
        if (!parsed.checkIns) parsed.checkIns = {};
        
        // Populate slots up to maxSlots
        for (let i = 1; i <= this.maxSlots; i++) {
          if (!parsed.slots[i]) {
            parsed.slots[i] = { number: i };
          }
        }
        return parsed;
      }
    } catch (e) {
      console.error("Error reading storage file, initializing fresh state", e);
    }

    // Initialize fresh state
    const slots: Record<number, Slot> = {};
    for (let i = 1; i <= this.maxSlots; i++) {
      slots[i] = { number: i };
    }
    return {
      users: {},
      slots: slots,
      checkIns: {}
    };
  }

  private saveState(): void {
    try {
      fs.writeFileSync(this.filePath, JSON.stringify(this.state, null, 2), 'utf-8');
    } catch (e) {
      console.error("Error writing storage file", e);
    }
  }

  async getUser(userId: number): Promise<TeleUser | null> {
    return this.state.users[userId] || null;
  }

  async saveUser(user: TeleUser): Promise<void> {
    this.state.users[user.id] = {
      ...this.state.users[user.id],
      ...user
    };
    this.saveState();
  }

  async chooseSlot(
    userId: number, 
    username: string | undefined, 
    firstName: string, 
    lastName: string | undefined, 
    slotNumber: number
  ): Promise<{ success: boolean; error?: string }> {
    if (slotNumber < 1 || slotNumber > this.maxSlots) {
      return { success: false, error: `Номер должен быть в пределах от 1 до ${this.maxSlots}!` };
    }

    const user = this.state.users[userId];
    if (user && user.slotNumber !== undefined) {
      return { success: false, error: `Вы уже выбрали номер — это Пидараз ${user.slotNumber}. Изменить его нельзя!` };
    }

    const slot = this.state.slots[slotNumber];
    if (slot && slot.userId !== undefined && slot.userId !== userId) {
      return { success: false, error: `Номер ${slotNumber} уже занят другими пидаразом!` };
    }

    // Assign the slot
    const nowStr = new Date().toISOString();
    
    // Create/update user object
    const updatedUser: TeleUser = {
      id: userId,
      username: username,
      first_name: firstName,
      last_name: lastName,
      slotNumber: slotNumber,
      registeredAt: nowStr
    };

    this.state.users[userId] = updatedUser;
    this.state.slots[slotNumber] = {
      number: slotNumber,
      userId: userId,
      occupiedAt: nowStr
    };

    this.saveState();
    return { success: true };
  }

  async getSlots(limit = 100): Promise<Slot[]> {
    const slots: Slot[] = [];
    for (let i = 1; i <= limit; i++) {
      slots.push(this.state.slots[i] || { number: i });
    }
    return slots;
  }

  async getAllUsers(): Promise<TeleUser[]> {
    return Object.values(this.state.users);
  }

  async recordCheckIn(userId: number, dateStr: string): Promise<boolean> {
    const user = this.state.users[userId];
    if (!user) return false;

    if (!this.state.checkIns[dateStr]) {
      this.state.checkIns[dateStr] = [];
    }

    if (!this.state.checkIns[dateStr].includes(userId)) {
      this.state.checkIns[dateStr].push(userId);
      user.lastCheckIn = dateStr;
      this.saveState();
      return true;
    }

    return false;
  }

  async getCheckIns(dateStr: string): Promise<number[]> {
    return this.state.checkIns[dateStr] || [];
  }
}
